geometric_kernels/sampling/probability_densities.py

```diff
@@ -107,21 +107,27 @@ def prop16(key, alpha, lengthscale):
     :param alpha: unnormalized coefficients of the mixture.
     :param lengthscale: length scale (kappa).
 
-    TODO: labify.
-    TODO: maybe rename this or integrate into the rejection sampler function.
     TODO: reparameterization trick.
     """
-    assert(len(alpha.shape) == 1)
-    m = len(alpha)-1
-    assert(m >= 0)
-    js = np.arange(0, m+1)
-    beta = np.power(2, (1-js)/2)/gamma_function((js+1)/2)*lengthscale
-    cs_unnorm = alpha/beta
-    cs = cs_unnorm/np.sum(cs_unnorm)
-    ind = np.random.choice(np.arange(0, m+1), 1, p=cs)[0]
-    return key, np.sqrt(np.random.chisquare(ind+1))/lengthscale
-
-def prop17(key, alpha, lengthscale, nu, dim):
+    assert B.rank(alpha) == 1
+    m = B.shape(alpha)[0] - 1
+    assert m >= 0
+    dtype = B.dtype(lengthscale)
+    js = B.range(dtype, 0, m + 1)
+
+    # Gamma((js+1)/2) should be positive real, so G = exp(log(abs(G)))
+    beta = 2 ** ((1 - js) / 2) / B.exp(B.loggamma((js + 1) / 2)) * lengthscale
+    cs_unnorm = alpha / beta
+    cs = cs_unnorm / B.sum(cs_unnorm)
+    key, ind = B.choice(key, js, 1, p=cs)
+
+    # Gamma(2nu, 2) distribution is the same as chi2(nu) distribution
```

Comment:
`2nu` -> `nu/2`, right? The code below implements the latter, wiki also supports the latter.